src/librustc_codegen_llvm/back/lto.rs

```diff
@@ -111,21 +110,35 @@ fn prepare_lto(
             }
 
             let archive = ArchiveRO::open(&path).expect("wanted an rlib");
-            let bytecodes = archive
+            let object_files = archive
                 .iter()
                 .filter_map(|child| child.ok().and_then(|c| c.name().map(|name| (name, c))))
-                .filter(|&(name, _)| name.ends_with(RLIB_BYTECODE_EXTENSION));
-            for (name, data) in bytecodes {
+                .filter(|&(name, _)| name.ends_with(".o"));
+            for (name, data) in object_files {
+                let name: &str = name;
+                let data: llvm::archive_ro::Child<'_> = data;
                 let _timer =
                     cgcx.prof.generic_activity_with_arg("LLVM_lto_load_upstream_bitcode", name);
-                info!("adding bytecode {}", name);
-                let bc_encoded = data.data();
+                info!("adding bitcode from {}", name);
+                let object_data: &[u8] = data.data();
 
-                let (bc, id) = match DecodedBytecode::new(bc_encoded) {
-                    Ok(b) => Ok((b.bytecode(), b.identifier().to_string())),
+                // njn: start by assuming fat object here, see what works and
+                // doesn't work
+
+                // njn: where does the id come from now? what exactly is this
+                // id? just use 'name' from above?
```

Comment:
`id` is the identifier we assigned the LLVM module which is needed for weird reasons. It has a bunch to do with how LLVM deduplicates things and registers modules, especially for ThinLTO. I don't think the identifier itself matters really so long as everything has a unique identifier. (which begs the question as to why identifiers are used anyway but alas!)

Reply:
So reading through this. I *think* that this is the same as the file stem without the `.o` on the end. It'd be good to confirm that though. Looks like we previously encoded both the bytecode and the identifier into the `*.bc.z` file, so we'll need to get the identifier from something. The `&module.name[..]` was written out, and again I *think* that's the same as the filename (or at least very closely correlated to it), but it would be good to double check.

In any case I think once the identifier is learned here (and it has to be calculated here independent of the contents of the bitcode) I think this is all looking good!

Reply:
Here are some sample `(name, id)` pairs from the existing code, which still uses `.bc.z` files:
```
(std-41707e39573ec24c.10qlq9isnkfvonab.rcgu.bc.z, 10qlq9isnkfvonab)
(panic_unwind-fb54cae34407c225.16je7xxussvyp7dt.rcgu.bc.z, 16je7xxussvyp7dt)
(hashbrown-fdaf2869972389f3.hashbrown.4xv4kx6v-cgu.0.rcgu.bc.z, hashbrown.4xv4kx6v-cgu.0)
(backtrace-e4bfa7af20dc2d47.backtrace.16m3lrxn-cgu.0.rcgu.bc.z, backtrace.16m3lrxn-cgu.0)
(rustc_demangle-dd74223c30447893.rustc_demangle.e6juhbfy-cgu.0.rcgu.bc.z, rustc_demangle.e6juhbfy-cgu.0)
(libc-e7cab63d4d6eee2e.libc.6yvkvw8t-cgu.0.rcgu.bc.z, libc.6yvkvw8t-cgu.0)
(alloc-04a6a115666ed9f6.14ru03yhgpmmgej5.rcgu.bc.z, 14ru03yhgpmmgej5)
(core-0560efb79be7b315.11guwpoxde9a3jp1.rcgu.bc.z, 11guwpoxde9a3jp1)
(core-0560efb79be7b315.8hsyj5o9fxj5ipu.rcgu.bc.z, 8hsyj5o9fxj5ipu)
```
So the pattern for `name` is clearly `$CRATE-$HASH.$ID.rcgu.bc.z`. `$ID` always contains a hash (though the length appears to vary) and sometimes has the crate name followed by `.` and sometimes has a `-cgu.0` suffix, which is a bit weird.

It seems best to try to reproduce the existing `id` exactly. One option is to do some string manipulation on the `name` to extract the `id`. Another option is to find the code that computes `name` and reuse it.

Reply:
Hm odd!

I'm not really sure why the module names are differing here. My best guess is it has to do with Cargo flags? All the "path" crates here have just the hash, but upstream "crates.io crates" have the crate name in them.

In any case I think it's actually safe to not reproduce this logic exactly. I think honestly you can probably take the literal filename and use it as the ID. The filename's always got the hash in it, and I think that the filename being the ID is roughly what C compilers/linkers/etc do for LTO anyway.

Reply:
Using `name` as is makes sense and is easy. I will try that.

Reply:
> (though the length appears to vary) and sometimes has the crate name followed by . and sometimes has a -cgu.0 suffi

I think it just skips printing leading zeros.